File: src/code-samples/langchain/long-term-memory-write-tool-postgres.py
# :snippet-start: long-term-memory-write-tool-postgres-py
from dataclasses import dataclass

from langchain.agents import create_agent
from langchain.tools import ToolRuntime, tool
from langchain_core.runnables import Runnable
from langgraph.store.postgres import PostgresStore  # type: ignore[import-not-found]
from typing_extensions import TypedDict


@dataclass
class Context:
    user_id: str


class UserInfo(TypedDict):
    name: str


@tool
def save_user_info(user_info: UserInfo, runtime: ToolRuntime[Context]) -> str:
    """Save user info."""
    assert runtime.store is not None
    runtime.store.put(("users",), runtime.context.user_id, dict(user_info))
    return "Successfully saved user info."


DB_URI = "postgresql://postgres:postgres@localhost:5432/postgres?sslmode=disable"
# :remove-start:
import os
import sys
from pathlib import Path

os.environ.setdefault("ANTHROPIC_API_KEY", "sk-ant-test-key")
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from conftest import get_postgres_uri, prepare_postgres_store

DB_URI = get_postgres_uri()
prepare_postgres_store(DB_URI)
# :remove-end:

with PostgresStore.from_conn_string(DB_URI) as store:
    store.setup()
    agent: Runnable = create_agent(
        "claude-sonnet-4-6",
        tools=[save_user_info],
        store=store,
        context_schema=Context,
    )

    agent.invoke(
        {"messages": [{"role": "user", "content": "My name is John Smith"}]},
        context=Context(user_id="user_123"),
    )
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    with PostgresStore.from_conn_string(DB_URI) as store:
        store.setup()
        store.put(("users",), "user_123", {"name": "John Smith"})
        saved_data = store.get(("users",), "user_123")
        assert saved_data is not None
        assert saved_data.value["name"] == "John Smith"
    print("✓ Write tool with PostgresStore works correctly")
# :remove-end:

File: src/code-samples/langchain/long-term-memory-write-tool-postgres.ts
// :snippet-start: long-term-memory-write-tool-postgres-js
import * as z from "zod";
import { tool, createAgent, type ToolRuntime } from "langchain";
import { PostgresStore } from "@langchain/langgraph-checkpoint-postgres/store";

const DB_URI =
  process.env.POSTGRES_URI ??
  "postgresql://postgres:postgres@localhost:5432/postgres?sslmode=disable";
const store = PostgresStore.fromConnString(DB_URI);
// :remove-start:
// Drop tables from prior runs (Python samples use different schema; CI shares one DB)
await (
  store as { core: { pool: { query: (q: string) => Promise<unknown> } } }
).core.pool.query(
  "DROP TABLE IF EXISTS public.store_vectors CASCADE; DROP TABLE IF EXISTS public.store CASCADE; DROP TABLE IF EXISTS public.store_migrations CASCADE;",
);
// :remove-end:
await store.setup();

const contextSchema = z.object({ userId: z.string() });

const UserInfo = z.object({ name: z.string() });

const saveUserInfo = tool(
  async (
    userInfo: z.infer<typeof UserInfo>,
    runtime: ToolRuntime<unknown, z.infer<typeof contextSchema>>,
  ) => {
    const userId = runtime.context.userId;
    if (!userId) throw new Error("userId is required");
    await runtime.store.put(["users"], userId, userInfo);
    return "Successfully saved user info.";
  },
  { name: "save_user_info", description: "Save user info", schema: UserInfo },
);

const agent = createAgent({
  model: "claude-sonnet-4-6",
  tools: [saveUserInfo],
  contextSchema,
  store,
});

await agent.invoke(
  { messages: [{ role: "user", content: "My name is John Smith" }] },
  { context: { userId: "user_123" } },
);

const result = await store.get(["users"], "user_123");
console.log(result?.value);
// :snippet-end:

// :remove-start:
async function main() {
  const DB_URI =
    process.env.POSTGRES_URI ||
    "postgresql://postgres:postgres@localhost:5432/postgres?sslmode=disable";
  const store = PostgresStore.fromConnString(DB_URI);

  try {
    await store.setup();

    const contextSchema = z.object({ userId: z.string() });
    const UserInfo = z.object({ name: z.string() });

    const saveUserInfo = tool(
      async (
        userInfo: z.infer<typeof UserInfo>,
        runtime: ToolRuntime<unknown, z.infer<typeof contextSchema>>,
      ) => {
        const userId = runtime.context.userId;
        if (!userId) throw new Error("userId is required");
        await runtime.store.put(["users"], userId, userInfo);
        return "Successfully saved user info.";
      },
      {
        name: "save_user_info",
        description: "Save user info",
        schema: UserInfo,
      },
    );

    // Test the tool directly - pass context and store in config (same shape as ToolRuntime)
    const saveResult = await saveUserInfo.invoke(
      { name: "John Smith" },
      { context: { userId: "user_123" }, store },
    );

    if (saveResult !== "Successfully saved user info.") {
      throw new Error("Expected save to succeed");
    }

    // Verify data was saved
    const savedData = await store.get(["users"], "user_123");
    if (!savedData) {
      throw new Error("Expected data to be saved");
    }
    if (savedData.value["name"] !== "John Smith") {
      throw new Error('Expected name to be "John Smith"');
    }

    console.log("✓ Write tool with PostgresStore works correctly");
  } finally {
    await store.stop();
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error(error);
    process.exit(1);
  });
}
// :remove-end:
